### ISSUE TITLE
blamm scan uses all cores

### ISSUE THREAD

**Issue by martindemko** [biointec/blamm#1]
Hi,
I specified the number of cores with parameter (-t 10) but it uses all 110 cores anyway, I tried even the long version of the parameter but doesn't change anything. The tool informs me that it is using 10 cores but that is not what I see in htop.
Can you help me?

Thanks,
Martin

**Comment by biointec** [comment]
Thank you for reaching out and thank you for your interest in ``blamm``.

Some BLAS vendors enable multithreading by default and this is probably what you are seeing. What BLAS implementation are you using?  If you are using Intel MKL, you can easily disable multithreading by setting the MKL_NUM_THREADS environment flag. For example, in Linux:

``export MKL_NUM_THREADS=1``

If this doesn't work, please let me know your OS/version and BLAS implementation.

**Comment by martindemko** [comment]
Thanks for such a quick response. I see now what could be the problem. Unfortunately, I installed MKL as well as OpenBLAS, both through CONDA into a dedicated environment so I actually don't know which one is used, to be honest, but I tried MKL_NUM_THREADS and also OPENBLAS_NUM_THREADS to set up to 10 cores but without effect. 

#uname -a:
Linux production 5.4.0-80-generic #90-Ubuntu SMP Fri Jul 9 22:49:44 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

Release:
Ubuntu 20.04.1 LTS

Conda env looks like this:
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                      1_llvm    conda-forge
alsa-lib                  1.2.3                h516909a_0    conda-forge
atk-1.0                   2.36.0               h3371d22_4    conda-forge
bedtools                  2.30.0               h7d7f7ad_2    bioconda
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.17.2               h7f98852_0    conda-forge
ca-certificates           2021.5.30            ha878542_0    conda-forge
cairo                     1.16.0            h6cf1ce9_1008    conda-forge
chrpath                   0.16              h7f98852_1002    conda-forge
cmake                     3.21.2               h8897547_0    conda-forge
dbus                      1.13.18              hb2f20db_0    anaconda
expat                     2.4.1                h9c3ff4c_0    conda-forge
font-ttf-dejavu-sans-mono 2.37                 hd3eb1b0_0  
font-ttf-inconsolata      3.000                h77eed37_0    conda-forge
font-ttf-source-code-pro  2.038                h77eed37_0    conda-forge
font-ttf-ubuntu           0.83                 hab24e00_0    conda-forge
fontconfig                2.13.1            hba837de_1005    conda-forge
fonts-conda-ecosystem     1                             0    conda-forge
fonts-conda-forge         1                             0    conda-forge
freetype                  2.10.4               h0708190_1    conda-forge
fribidi                   1.0.10               h516909a_0    conda-forge
gdk-pixbuf                2.42.6               h04a7f16_0    conda-forge
gettext                   0.21.0               hf68c758_0  
giflib                    5.2.1                h516909a_2    conda-forge
glib                      2.68.4               h9c3ff4c_0    conda-forge
glib-tools                2.68.4               h9c3ff4c_0    conda-forge
gnuplot                   5.4.1                hec6539f_2    conda-forge
graphite2                 1.3.14               h23475e2_0    anaconda
gst-plugins-base          1.18.5               hf529b03_0    conda-forge
gstreamer                 1.18.5               h76c114f_0    conda-forge
gtk2                      2.24.33              h539f30e_1    conda-forge
harfbuzz                  2.9.1                h83ec7ef_0    conda-forge
icu                       68.1                 h2531618_0  
jbig                      2.1               h7f98852_2003    conda-forge
jpeg                      9d                   h516909a_0    conda-forge
krb5                      1.19.2               hcc1bbae_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
lerc                      2.2.1                h2531618_0  
libclang                  11.1.0          default_ha53f305_1    conda-forge
libcurl                   7.78.0               h2574ce0_0    conda-forge
libdeflate                1.7                  h27cfd23_5  
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libevent                  2.1.10               hcdb4288_3    conda-forge
libffi                    3.3                  he6710b0_2    anaconda
libgcc-ng                 11.1.0               hc902ee8_8    conda-forge
libgd                     2.3.2                h78a0170_0    conda-forge
libgfortran               3.0.0                         1    conda-forge
libgfortran-ng            11.1.0               h69a702a_8    conda-forge
libgfortran5              11.1.0               h6c583b3_8    conda-forge
libglib                   2.68.4               h3e27bee_0    conda-forge
libiconv                  1.16                 h516909a_0    conda-forge
libllvm11                 11.1.0               hf817b99_2    conda-forge
libnghttp2                1.43.0               h812cca2_0    conda-forge
libogg                    1.3.5                h27cfd23_1  
libopenblas               0.3.17          pthreads_h8fe5266_1    conda-forge
libopus                   1.3.1                h7f98852_1    conda-forge
libpng                    1.6.37               hed695b0_2    conda-forge
libpq                     13.3                 hd57d9b9_0    conda-forge
libssh2                   1.10.0               ha56f1ee_0    conda-forge
libstdcxx-ng              11.1.0               h56837e0_8    conda-forge
libtiff                   4.3.0                hf544144_1    conda-forge
libuuid                   2.32.1            h14c3975_1000    conda-forge
libuv                     1.42.0               h7f98852_0    conda-forge
libvorbis                 1.3.7                he1b5a44_0    conda-forge
libwebp                   1.2.1                h3452ae3_0    conda-forge
libwebp-base              1.2.1                h7f98852_0    conda-forge
libxcb                    1.14                 h7b6447c_0    anaconda
libxkbcommon              1.0.3                he3ba5ed_0    conda-forge
libxml2                   2.9.12               h72842e0_0    conda-forge
llvm-openmp               12.0.1               h4bd325d_1    conda-forge
lz4-c                     1.9.3                h9c3ff4c_1    conda-forge
mkl                       2021.3.0           h726a3e6_557    conda-forge
mysql-common              8.0.25               ha770c72_2    conda-forge
mysql-libs                8.0.25               hfa10184_2    conda-forge
ncurses                   6.2                  h58526e2_4    conda-forge
nspr                      4.30                 h9c3ff4c_0    conda-forge
nss                       3.69                 hb5efdd6_0    conda-forge
openblas                  0.3.17          pthreads_h4748800_1    conda-forge
openssl                   1.1.1l               h7f98852_0    conda-forge
pango                     1.48.9               hb8ff022_0    conda-forge
pcre                      8.45                 h295c915_0  
pixman                    0.40.0               h36c2ea0_0    conda-forge
python                    3.9.7           h49503c6_0_cpython    conda-forge
qt                        5.12.9               hda022c4_4    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
rhash                     1.4.1                h7f98852_0    conda-forge
sqlite                    3.36.0               hc218d9a_0  
tbb                       2021.3.0             h4bd325d_0    conda-forge
tk                        8.6.11               h27826a3_1    conda-forge
tzdata                    2021a                he74cb21_1    conda-forge
xorg-kbproto              1.0.7             h14c3975_1002    conda-forge
xorg-libice               1.0.10               h516909a_0    conda-forge
xorg-libsm                1.2.3             hd9c2040_1000    conda-forge
xorg-libx11               1.7.2                h7f98852_0    conda-forge
xorg-libxext              1.3.4                h7f98852_1    conda-forge
xorg-libxrender           0.9.10            h7f98852_1003    conda-forge
xorg-libxt                1.2.1                h7f98852_2    conda-forge
xorg-renderproto          0.11.1            h14c3975_1002    conda-forge
xorg-xextproto            7.3.0             h14c3975_1002    conda-forge
xorg-xproto               7.0.31            h27cfd23_1007  
xz                        5.2.5                h516909a_1    conda-forge
zlib                      1.2.11            h516909a_1010    conda-forge
zstd                      1.5.0                ha95c52a_0    conda-forge

**Comment by biointec** [comment]
Let us first try and figure out whether you use MKL or OpenBLAS. I assume you compiled blamm using CMake. When you configure CMake, it tells you what BLAS library was found.  For example, assuming the ``build`` subdirectory to build blamm:

```bash
mkdir build
cd build
cmake ..
```

produces at some point on my system:

``-- Found BLAS: /usr/lib/x86_64-linux-gnu/libblas.so``

You can double-check this during the compilation of blamm as well:

``make VERBOSE=1``

informs me right at the end:

```bash
[100%] Linking CXX executable blamm
/usr/bin/cmake -E cmake_link_script CMakeFiles/blamm.dir/link.txt --verbose=1
/usr/bin/c++  -Wall -Wextra -std=c++11 -O3  -rdynamic CMakeFiles/blamm.dir/src/dict.cpp.o CMakeFiles/blamm.dir/src/hist.cpp.o CMakeFiles/blamm.dir/src/blstools.cpp.o CMakeFiles/blamm.dir/src/bls.cpp.o CMakeFiles/blamm.dir/src/phylotree.cpp.o CMakeFiles/blamm.dir/src/matrix.cpp.o CMakeFiles/blamm.dir/src/pwmscan.cpp.o CMakeFiles/blamm.dir/src/motif.cpp.o CMakeFiles/blamm.dir/src/sequence.cpp.o CMakeFiles/blamm.dir/src/ortho.cpp.o CMakeFiles/blamm.dir/src/species.cpp.o CMakeFiles/blamm.dir/src/settings.cpp.o  -o blamm  -lblas -lpthread
```
Indeed, the ``-lblas`` means I am linking against the Ubuntu built-in `libblas` library.

Can you provide me with your output?

**Comment by biointec** [comment]
For some BLAS implementations, you can specify the number of threads with the `OMP_NUM_THREADS` enviroment variable.  So:

``export OMP_NUM_THREADS=1``

is something you can try.

**Comment by martindemko** [comment]
Hi,
so I checked and I use MKL indeed. Anyway, I figured it out with your help of course. I just tried to set up MKL_NUM_THREADS to 1 and kept 10 threads in blamm command and it really took just 10 cores. When I did it vice versa (MKL_NUM_THREADS=10, -t 1) it used 10 cores again. So I tried to set MKL_NUM_THREADS=2 and it used 20 cores from which I assume these two settings are multiplied to get a resulting number of cores. This is non-intuitive to me but still at least understandable. So I know now what to do next.
Thanks for your quick help.

Martin

**Comment by jfostier** [comment]
You are absolutely right. Specifying the number of threads in `blamm` controls how many concurrent matrix-matrix multiplications are issued.  The `MKL_NUM_THREADS` variable controls how many threads are used *per* matrix-matrix multiplication. For optimal performance, set the latter to 1 and simply control the number of threads using `blamm`.  I have updated the manual accordingly. Closing this issue.
